synapse/handlers/register.py

```diff
@@ -627,8 +634,24 @@ def _register_with_store(self, user_id, token=None, password_hash=None,
                 create_profile_with_displayname=create_profile_with_displayname,
                 admin=admin,
                 user_type=user_type,
+                address=address,
             )
         else:
+            # Don't rate limit for app services
+            if appservice_id is None and address is not None:
+                time_now = self.clock.time()
+
+                allowed, time_allowed = self.ratelimiter.send_message(
+                    address, time_now_s=time_now,
+                    msg_rate_hz=self.hs.config.rc_auth_requests_per_second,
+                    burst_count=self.hs.config.rc_auth_request_burst_count,
+                )
+
+                if not allowed:
+                    raise LimitExceededError(
+                        retry_after_ms=int(1000 * (time_allowed - time_now)),
+                    )
```

Comment:
Thinking about this, I think we should probably have this outside of the `if`, as then the limiting will happen on both master and worker, which is nice because:

1. Doing the update on the worker means that the initial check in `on_POST` works (it won't if the update only happens on master)
2. Doing the update on master means that if we do end up having multiple workers performing registration, the ratelimiting is correctly applied across all of them.

Reply:
Right, makes sense.